nilearn/regions/region_extractor.py

```diff
@@ -0,0 +1,320 @@
+        raise ValueError(message)
+
+    if mask_img is not None:
+        if not _check_same_fov(maps_img, mask_img):
+            mask_img = resample_img(mask_img,
+                                    target_affine=maps_img.get_affine(),
+                                    target_shape=maps_img.shape[:3],
+                                    interpolation="nearest")
+            mask_data, _ = masking._load_mask_img(mask_img)
+            # Set as 0 to the values which are outside of the mask
+            maps[mask_data == 0.] = 0.
+
+    for index in range(maps.shape[-1]):
+        regions = []
+        map_3d = maps[..., index]
+        # Mark the seeds using random walker
+        if extract_type == 'local_regions':
+            smooth_map = _smooth_array(map_3d, affine=affine, fwhm=smoothing_fwhm)
+            seeds = _peak_local_max(smooth_map)
+            seeds_label, seeds_id = label(seeds)
+            # Assign -1 to values which are 0. to indicate to ignore
+            seeds_label[map_3d == 0.] = -1
+            rw_maps = _random_walker(map_3d, seeds_label)
+            # Now simply replace "-1" with "0" for regions seperation
+            rw_maps[rw_maps == -1] = 0.
+            label_maps = rw_maps
+        else:
+            # Connected component extraction
+            label_maps, n_labels = label(map_3d)
+
+        # Takes the size of each labelized region data
+        labels_size = np.bincount(label_maps.ravel())
+        # set background labels sitting in zero index to zero
+        labels_size[0] = 0.
+        for label_id, label_size in enumerate(labels_size):
+            if label_size > min_region_size:
+                region_data = (label_maps == label_id) * map_3d
+                region_img = new_img_like(maps_img, region_data)
+                regions.append(region_img)
+
+        index_of_each_map.extend([index] * len(regions))
+        all_regions_imgs.extend(regions)
+
+    regions_extracted_img = concat_niimgs(all_regions_imgs)
+
+    return regions_extracted_img, index_of_each_map
+
+
+class RegionExtractor(NiftiMapsMasker):
+    """ Class for brain region extraction.
+
+    Region Extraction is a post processing technique which
+    is implemented to automatically segment each brain atlas maps
+    into different set of separated brain activated region.
+    Particularly, to show that each decomposed brain maps can be
+    used to focus on a target specific Regions of Interest analysis.
+
+    Parameters
+    ----------
+    maps_img: 4D Niimg-like object
+       Image containing a set of whole brain atlas maps or statistically
+       decomposed brain maps.
+
+    mask_img: Niimg-like object or None,  default None, optional
+        Mask to be applied to input data, passed to NiftiMapsMasker.
+        If None, no masking is applied.
+
+    min_region_size: int, default 50, optional
+        Minimum number of voxels for a region to be kept. Useful to suppress
+        small spurious regions.
```

Comment:
In theory, you're completely right. 
In practice, I'm not so sure: people are really not used to thinking in terms of 'I expect brain areas of size at least 200mm^3', and actually more easily think in terms of voxel counts. This is obvious when you look at the SPM interface.